winpr/libwinpr/sspi/Kerberos/kerberos.c

```diff
@@ -486,106 +453,448 @@ static SECURITY_STATUS SEC_ENTRY kerberos_QueryCredentialsAttributesA(PCredHandl
+
+			/* Set auth_context options */
+			if ((rv = krb5_auth_con_init(context->ctx, &context->auth_ctx)))
+				goto cleanup;
+			if ((rv = krb5_auth_con_setflags(context->ctx, context->auth_ctx,
+			                                 KRB5_AUTH_CONTEXT_DO_SEQUENCE |
+			                                     KRB5_AUTH_CONTEXT_USE_SUBKEY)))
+				goto cleanup;
+			if ((rv = krb5_auth_con_set_req_cksumtype(context->ctx, context->auth_ctx,
+			                                          GSS_CHECKSUM_TYPE)))
+				goto cleanup;
+
+			/* Write the checksum (channel binding and delegation not implemented) */
+			Data_Write_UINT32(cksum_contents, 16);
+			Data_Write_UINT32((cksum_contents + 20), context->flags);
+
+			/* Get a service ticket */
+			if ((rv = krb5_sname_to_principal(context->ctx, host, sname, KRB5_NT_SRV_HST,
+			                                  &in_creds.server)))
+				goto cleanup;
+
+			if ((rv = krb5_cc_get_principal(context->ctx, credentials->ccache, &in_creds.client)))
+				goto cleanup;
+
+			if ((rv = krb5_get_credentials(context->ctx, context->u2u ? KRB5_GC_USER_USER : 0,
+			                               credentials->ccache, &in_creds, &creds)))
+				goto cleanup;
+
+			/* Make the AP_REQ message */
+			if ((rv = krb5_mk_req_extended(context->ctx, &context->auth_ctx, ap_flags, &cksum,
+			                               creds, &output_token)))
+				goto cleanup;
+
+			if (!sspi_gss_wrap_token(output_buffer,
+			                         context->u2u ? &kerberos_u2u_OID : &kerberos_OID,
+			                         TOK_ID_AP_REQ, &output_token))
+				goto cleanup;
+
+			if (context->flags & SSPI_GSS_C_SEQUENCE_FLAG)
+			{
+				krb5_auth_con_getlocalseqnumber(context->ctx, context->auth_ctx,
+				                                (INT32*)&context->local_seq);
+				context->remote_seq ^= context->local_seq;
+			}
+
+			krb5_auth_con_getkey_k(context->ctx, context->auth_ctx, &context->session_key);
+			krb5_auth_con_getsendsubkey_k(context->ctx, context->auth_ctx, &context->initiator_key);
+
+			context->state = KERBEROS_STATE_AP_REP;
+
+			break;
+
+		case KERBEROS_STATE_AP_REP:
+
+			if (tok_id == TOK_ID_AP_REP)
+			{
+				if ((rv = krb5_rd_rep(context->ctx, context->auth_ctx, &input_token, &reply)))
+					goto cleanup;
+				krb5_free_ap_rep_enc_part(context->ctx, reply);
+			}
+			else if (tok_id == TOK_ID_ERROR)
+			{
+				if (!(rv = krb5_rd_error(context->ctx, &input_token, &error)))
+				{
+					WLog_ERR(TAG, "KRB_ERROR: %s", error->text.data);
+					krb5_free_error(context->ctx, error);
+				}
+				goto cleanup;
+			}
+			else
+				goto bad_token;
+
+			if (context->flags & SSPI_GSS_C_SEQUENCE_FLAG)
+				krb5_auth_con_getremoteseqnumber(context->ctx, context->auth_ctx,
+				                                 (INT32*)&context->remote_seq);
+
+			krb5_auth_con_getrecvsubkey_k(context->ctx, context->auth_ctx, &context->acceptor_key);
+
+			context->state = KERBEROS_STATE_FINAL;
+
+			break;
+
+		case KERBEROS_STATE_FINAL:
```

Comment:
Not sure but maybe a missing break (otherwise STATE_FINAL will abort with an error and never execute the STATE_FINAL block below)?

Reply:
We set the state to FINAL after processing AP_REP - that's when the later block executes. To receive a token after we already processed the AP_REP would be invalid.

Reply:
@fifthdegree thx for clarification